vault/seal/transit/transit.go

```diff
@@ -0,0 +1,233 @@
+)
+
+// Seal is a seal that leverages Vault's Transit secret
+// engine
+type Seal struct {
+	logger  log.Logger
+	client  *api.Client
+	renewer *api.Renewer
+
+	mountPath string
+	keyName   string
+
+	currentKeyID *atomic.Value
+}
+
+var _ seal.Access = (*Seal)(nil)
+
+// NewSeal creates a new transit seal
+func NewSeal(logger log.Logger) *Seal {
+	s := &Seal{
+		logger:       logger.ResetNamed("seal-transit"),
+		currentKeyID: new(atomic.Value),
+	}
+	s.currentKeyID.Store("")
+	return s
+}
+
+// SetConfig processes the config info from the server config
+func (s *Seal) SetConfig(config map[string]string) (map[string]string, error) {
+	if config == nil {
+		config = map[string]string{}
+	}
+
+	switch {
+	case os.Getenv("VAULT_TRANSIT_SEAL_MOUNT_PATH") != "":
+		s.mountPath = os.Getenv("VAULT_TRANSIT_SEAL_MOUNT_PATH")
+	case config["mount_path"] != "":
+		s.mountPath = config["mount_path"]
+	default:
+		return nil, fmt.Errorf("mount_path is required")
+	}
+
+	switch {
+	case os.Getenv("VAULT_TRANSIT_SEAL_KEY_NAME") != "":
+		s.keyName = os.Getenv("VAULT_TRANSIT_SEAL_KEY_NAME")
+	case config["key_name"] != "":
+		s.keyName = config["key_name"]
+	default:
+		return nil, fmt.Errorf("key_name is required")
+	}
+
+	var disableRenewal bool
+	var disableRenewalRaw string
+	switch {
+	case os.Getenv("VAULT_TRANSIT_SEAL_DISABLE_RENEWAL") != "":
+		disableRenewalRaw = os.Getenv("VAULT_TRANSIT_SEAL_DISABLE_RENEWAL")
+	case config["disable_renewal"] != "":
+		disableRenewalRaw = config["disable_renewal"]
+	}
+	if disableRenewalRaw != "" {
+		var err error
+		disableRenewal, err = strconv.ParseBool(disableRenewalRaw)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var namespace string
+	switch {
+	case os.Getenv("VAULT_NAMESPACE") != "":
+		namespace = os.Getenv("VAULT_NAMESPACE")
+	case config["namespace"] != "":
+		namespace = config["namespace"]
+	}
+
+	apiConfig := api.DefaultConfig()
+	if config["address"] != "" {
+		apiConfig.Address = config["address"]
+	}
+
+	if s.client == nil {
+		client, err := api.NewClient(apiConfig)
+		if err != nil {
+			return nil, err
+		}
+		if config["token"] != "" {
+			client.SetToken(config["token"])
+		}
+		if namespace != "" {
+			client.SetNamespace(namespace)
+		}
+		if client.Token() == "" {
+			return nil, errors.New("missing token")
+		}
+		s.client = client
+
+		if !disableRenewal {
+			// Renew the token immediately to get a secret to pass to renewer
+			secret, err := client.Auth().Token().RenewTokenAsSelf(s.client.Token(), 0)
+			// If we don't get an error renewing, set up a renewer.  The token may not be renewable or not have
+			// permission to renew-self.
+			if err == nil {
+				renewer, err := s.client.NewRenewer(&api.RenewerInput{
+					Secret: secret,
+				})
+				if err != nil {
+					return nil, err
+				}
+				s.renewer = renewer
+
+				go func() {
+					for {
+						select {
+						case err := <-renewer.DoneCh():
+							s.logger.Info("shutting down token renewal")
```

Comment:
So i wonder what we should do if we lose the token, if sealwrap is enabled we'd stop being able to return any wrapped value?

Reply:
Not really sure.  This really isn't much different if someone changes the creds on any autoseal, it just stops working sorta.